### ISSUE TITLE
Eventlet 0.18.0 and 0.18.1 are no longer on pypi

### ISSUE THREAD

**Issue by reversefold** [eventlet/eventlet#298]

**Comment by manasdk** [comment]
Removing old packages from pypi leads to issues when packages version are pinned. e.g. https://github.com/StackStorm/mistral/blob/st2-1.3.1/requirements.txt#L32 we pin to a specific version at time of release to avoid issue due to upstream failures. 

We have a CI process in place that performs this pinning and not having the package on pipit leads to a failure during installs.

Is there a strong reason to delete old patch versions?

**Comment by temoto** [comment]
TL;DR: Yes, the reason is critical bug. It's not everyday practice.

You should be aware that CI against version range leads to inconsistency, that is undefined CI results.

0.18 line introduces new semantics for socket.send: it does not attempt to finish partial send. Previously, it would loop like socket.sendall. Now we're doing minimal work as per documented API. So it will not break proper applications.

On Python3 it's a problem, though. GreenFileIO inherits from stdlib IOBase which does no buffering. So if you have code like sock.makefile('w').writelines(data), you loose part of data after first EAGAIN, that could not be sent right away.

eventlet.wsgi is famous piece of art containing such pattern. 0.18.2 contains crutch fix in wsgi. General fix for any makefile().writelines() is coming soon.

See #295 for lots of details, emotional discussion below commit 5bf0a6f32b3e4459b38ad1895c9eb4b0b483dae1.

This issue will remain open until all related issues are sorted out.

**Comment by temoto** [comment]
related issue #296

**Comment by temoto** [comment]
New status update: `sock.makefile('w').writelines()` could only loose data if you explicitly disable file-object buffering with `makefile('w', 0)`. That is the case with `eventlet.wsgi` (and also any other WSGI server that uses stdlib `BaseHttpServer`). Default `makefile` object is buffered and does not loose data.
